Allow creation of tables with Encryption enabled at rest (#518)

leobarcellos opened the pull request:
https://github.com/pynamodb/PynamoDB/issues/518

encryption_enabled argument can be passed to Model.create_table
`Model.create_table(wait=True, encryption_enabled=True)`

leobarcellos commented:
Hi @garrettheel , I would love to, but I'm kinda new to Python, I don't have the necessary skill yet. Gonna study and add a test as soon as I can

garrettheel commented:
@leobarcellos no worries! Hopefully the existing tests should have some good examples, but just let me know if you need help with anything

garrettheel commented:
hey @leobarcellos, is there anything I can help with for those tests? thanks!

leobarcellos commented:
hi @garrettheel ! I added a test on test_base_connection.py.

I was trying to add a test in test_model.py using the info from describe_table to check, could not yet make work.

leobarcellos commented:
> hey @leobarcellos, is there anything I can help with for those tests? thanks!

> hi @garrettheel ! I added a test on test_base_connection.py.
> 
> I was trying to add a test in test_model.py using the info from describe_table to check, could not yet make work.

Just added a test in `test_model.py` also, using a updated `DESCRIBE_TABLE_DATA` from `tests/data.py`.

Please let me know if these tests are useful and if i'm going in the right direction.
Thank you!,

dmulter commented:
> #518
> 
> encryption_enabled argument can be passed to Model.create_table
> `Model.create_table(wait=True, encryption_enabled=True)`

It should be noted in new documentation describing this feature that setting `encryption_enabled=True` will enable encryption with a user-provided KMS key. Setting it to False, or not specifying it, will enable encryption with the AWS-provided KMS key. Unless you want to manage keys yourself, most people probably don't want to use the new feature.

leobarcellos commented:
> > #518
> > encryption_enabled argument can be passed to Model.create_table
> > `Model.create_table(wait=True, encryption_enabled=True)`
> 
> It should be noted in new documentation describing this feature that setting `encryption_enabled=True` will enable encryption with a user-provided KMS key. Setting it to False, or not specifying it, will enable encryption with the AWS-provided KMS key. Unless you want to manage keys yourself, most people probably don't want to use the new feature.

You are right! I didn't realize AWS did this. 

I'm going to update this PR with correct naming and docstring to prevent undesired use of KMS, which additional charges may apply.

Just updating because the job is already done, this aws update from november 2018 really makes this PR very specific.

Reference: [Amazon DynamoDB encrypts all customer data at rest](https://aws.amazon.com/about-aws/whats-new/2018/11/amazon-dynamodb-encrypts-all-customer-data-at-rest)

venkateshanbalagan commented:
Hi, I am trying to test this code to create a table with KMS key. While I see we can provide kms_encryption_enabled = True, but I dont see where we receive the actual KMS key to encrypt?. Am I missing something or this feature is yet to be added. Pls advice.

jack1902 commented:
Is this likely to be merged? As the lack of Kms encryption literally renders this library useless

ikonst commented:
@jack1902 Thanks for the feedback. I'm sorry this library is useless to you.

Generally speaking, provisioning AWS resources isn't the strong side of this library, as in significant deployments you would usually use an infrastructure management framework like Terraform or CloudFormation.

jack1902 commented:
Don't get me wrong, I've used this heavily in the past but for this use case I have to have encryption at rest. Ideally I'd go for Aws managed Kms keys for dynamodb but it doesn't seem like I can specify this for the table /model?

Am I right in assuming that if I created the table using terraform / cloudformation / console etc that I won't be able to read from the table using this library?

ikonst commented:
@jack1902 It's encryption-at-rest so it only changes how your data is stored to "durable media" -- basically, if someone breaks into an AWS datacenter and steals a hard drive, it'll be encrypted.

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/EncryptionAtRest.html

> When you access an encrypted table, DynamoDB decrypts the table data transparently. You can switch between the AWS owned CMK, AWS managed CMK, and customer managed CMK at any given time. You don't have to change any code or applications to use or manage encrypted tables. DynamoDB continues to deliver the same single-digit millisecond latency that you have come to expect, and all DynamoDB queries work seamlessly on your encrypted data.

brno32 commented:
Are there any plans to merge this? I'm looking to declare my table entirely within code. It'd be neat if the `Meta` class had some attribute for a KMS key arn. This library shouldn't provision a KMS key for you, but it'd be neat if you could point to one you already provisioned yourself.

LeoGrosjean commented:
what need to be done to merge it ?
I can help if needed
